File: cpp/BalancedBinaryTree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    int height(TreeNode *root) {
        if (root == NULL) 
            return 0;
        return max(height(root->left), height(root->right)) + 1;
    }
    
    bool isBalanced(TreeNode *root) {
        if (root == NULL)
            return true;
        int hl = height(root->left);
        int hr = height(root->right);
        if (abs(hl - hr) > 1)
            return false;
        return (isBalanced(root->left) && isBalanced(root->right));
    }
};

File: Algorithms/python/110-BalancedBinaryTree/BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        # we can use a simple solution to solve this problem
        if root is None:
            return True

        lh = Solution.get_tree_height(root.left)
        rh = Solution.get_tree_height(root.right)

        return abs(lh - rh) <= 1 and self.isBalanced(root.left) and self.isBalanced(root.right)

    @staticmethod
    def get_tree_height(root):
        if root is None:
            return 0
        return max(Solution.get_tree_height(root.left),
                   Solution.get_tree_height(root.right)) + 1
